packages/test-interface-criteria/src/criteria/TestOpportunityBookable.js

```diff
@@ -1,30 +1,49 @@
 const moment = require('moment');
 
-const CriteriaFutureScheduledOpportunity = require('./CriteriaFutureScheduledOpportunity')
+const { CriteriaFutureScheduledOpportunity } = require('./CriteriaFutureScheduledOpportunity');
+const { getRemainingCapacity, createCriteria } = require('./criteriaUtils');
 
-/*
-  Implements https://openactive.io/test-interface#TestOpportunityBookable
-*/
+/**
+ * @typedef {import('../types/Criteria').OpportunityConstraint} OpportunityConstraint
+ * @typedef {import('../types/Criteria').OfferConstraint} OfferConstraint
+ */
 
-module.exports = class TestOpportunityBookable extends CriteriaFutureScheduledOpportunity {
-  get opportunityConstraints() {
-    return {
-      ...super.opportunityConstraints,
-      'Remaining capacity must be non-zero': opportunity => this.getRemainingCapacity(opportunity) > 0
-    };
-  }
+/**
+ * @type {OpportunityConstraint}
+ */
+function remainingCapacityMustBeNonZero(opportunity) {
+  return getRemainingCapacity(opportunity) > 0;
+}
 
-  get offerConstraints() {
-    return {
-      ...super.offerConstraints,
-      'Must have "bookable" offer': (x, opportunity) =>
-        (Array.isArray(x.availableChannel) && x.availableChannel.includes("https://openactive.io/OpenBookingPrepayment"))
-        && x.advanceBooking != "https://openactive.io/Unavailable"
-        && (!x.validFromBeforeStartDate || moment(opportunity.startDate).subtract(moment.duration(x.validFromBeforeStartDate)).isBefore())
-    }
-  }
+/**
+ * @type {OfferConstraint}
+ */
+function mustHaveBookableOffer(offer, opportunity) {
+  return (Array.isArray(offer.availableChannel) && offer.availableChannel.includes('https://openactive.io/OpenBookingPrepayment'))
+    && offer.advanceBooking !== 'https://openactive.io/Unavailable'
+    && (!offer.validFromBeforeStartDate || moment(opportunity.startDate).subtract(moment.duration(offer.validFromBeforeStartDate)).isBefore());
+}
 
-  get name() {
-    return 'TestOpportunityBookable';
-  }
-}
\ No newline at end of file
+/**
+ * Implements https://openactive.io/test-interface#TestOpportunityBookable
+ */
+const TestOpportunityBookable = createCriteria(
+  'TestOpportunityBookable',
+  [
+    [
+      'Remaining capacity must be non-zero',
+      remainingCapacityMustBeNonZero,
+    ],
+  ],
+  [
+    [
+      'Must have "bookable" offer',
+      mustHaveBookableOffer,
+    ],
+  ],
+  CriteriaFutureScheduledOpportunity,
+);
```

Comment:
@lukehesluke on reflection, should createCriteria not accept an object so that we can label each parameter with keys (and still benefit from typing). Reflecting on the above, just by looking at this code it's not immediately obvious what's going on? Perhaps it would be more self-documenting this way?

Reply:
Yes @nickevansuk that would make sense. There are times when I feel as if almost all parameters (in JS, which does not support keyword args like many other languages) should be in an object unless they're extremely obvious from function position.

I expect the compiler can probably turn the object into a struct rather than a hashmap as well if it can see destructuring in the function signature, so there's no perf overhead from doing (almost) everything in that style.

The above is not too bad in an editor like VSCode, where you can see the arg names (and types) when you hover over the function but I completely agree that an object would make that clear from a moment's glance rather than having to hover